service/matching/version_rule_helpers.go

```diff
@@ -515,3 +539,154 @@ func calcRampThreshold(id string) float64 {
 	h := farm.Fingerprint32([]byte(id))
 	return 100 * (float64(h) / (float64(math.MaxUint32) + 1))
 }
+
+/*
+e.g.
+Redirect Rules:
+10
+^
+|
+1 <------ 2
+^
+|
+5 <------ 3 <------ 4
+
+Assignment Rules:
+[ (3, 50%), (2, nil) ]
+*/
+func getUpstreamBuildIds(
+	buildId string,
+	redirectRules []*taskqueue.TimestampedCompatibleBuildIdRedirectRule) []string {
+	var upstream []string
+	directSources := getSourcesForTarget(buildId, redirectRules)
+
+	for _, src := range directSources {
+		upstream = append(upstream, src)
+		upstream = append(upstream, getUpstreamBuildIds(src, redirectRules)...)
+	}
+
+	return upstream
+}
+
+// getSourcesForTarget gets the first-degree sources for any redirect rule targeting buildId
+func getSourcesForTarget(buildId string, redirectRules []*taskqueue.TimestampedCompatibleBuildIdRedirectRule) []string {
+	var sources []string
+	for _, rr := range redirectRules {
+		if rr.GetRule().GetTargetBuildId() == buildId {
+			sources = append(sources, rr.GetRule().GetSourceBuildId())
+		}
+	}
+
+	return sources
+}
+
+func existsBackloggedActivityOrWFAssignedToAny(ctx context.Context,
```

Comment:
not sure if we want to ship commented code. maybe we can just get rid of the backlog check for now and just add a TODO for it in getBuildIdTaskReachability.

Reply:
removed